src/widgets/maybe.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget for optional data, with different `Some` and `None` children.
+
+use druid::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, Size,
+    UpdateCtx, Widget, WidgetExt, WidgetPod,
+};
+
+use druid::widget::SizedBox;
+
+/// A widget that switches between two possible child views, for `Data` that
+/// is `Option<T>`.
+pub struct Maybe<T> {
+    some_maker: Box<dyn Fn() -> Box<dyn Widget<T>>>,
+    none_maker: Box<dyn Fn() -> Box<dyn Widget<()>>>,
+    widget: MaybeWidget<T>,
+}
+
+enum MaybeWidget<T> {
+    Some(WidgetPod<T, Box<dyn Widget<T>>>),
+    None(WidgetPod<(), Box<dyn Widget<()>>>),
+}
+
+impl<T: Data> Maybe<T> {
+    /// Create a new `Maybe` widget with a `Some` and a `None` branch.
+    pub fn new<W1, W2>(
+        // we make these generic so that the caller doesn't have to explicitly
+        // box. We don't technically *need* to box, but it seems simpler.
+        some_maker: impl Fn() -> W1 + 'static,
+        none_maker: impl Fn() -> W2 + 'static,
+    ) -> Maybe<T>
+    where
+        W1: Widget<T> + 'static,
+        W2: Widget<()> + 'static,
+    {
+        let widget = MaybeWidget::Some(WidgetPod::new(some_maker().boxed()));
+        Maybe {
+            some_maker: Box::new(move || some_maker().boxed()),
+            none_maker: Box::new(move || none_maker().boxed()),
+            widget,
+        }
+    }
+
+    /// Create a new `Maybe` widget where the `None` branch is an empty widget.
+    #[allow(dead_code)]
+    pub fn or_empty<W1: Widget<T> + 'static>(some_maker: impl Fn() -> W1 + 'static) -> Maybe<T> {
+        Self::new(some_maker, || SizedBox::empty())
+    }
```

Comment:
I wonder how `SizedBox::empty()` interacts with `Flex` and space between / around as it would probably cause duplicated spacing.
Might be interesting to skip zero sized children in `Flex`

Reply:
that's an interesting point!